components/script/dom/htmlimageelement.rs

```diff
@@ -271,106 +268,360 @@ impl HTMLImageElement {
-                            // Step 11, substep 5
-                            let img = self.img.root();
-                            img.current_request.borrow_mut().source_url = Some(self.src.into());
-                            img.upcast::<EventTarget>().fire_event(atom!("error"));
-                            img.upcast::<EventTarget>().fire_event(atom!("loadend"));
+                }
+            },
+            ImageRequestPhase::Current => {
+                let mut current_request = self.current_request.borrow_mut();
+                // step 12.4, create a new "image_request"
+                match (current_request.parsed_url.clone(), current_request.state) {
+                    (Some(parsed_url), State::PartiallyAvailable) => {
+                        // Step 12.2
+                        if parsed_url == *url {
+                            // 12.3 abort pending request
+                            let mut pending_request = self.pending_request.borrow_mut();
+                            pending_request.image = None;
+                            pending_request.parsed_url = None;
+                            LoadBlocker::terminate(&mut pending_request.blocker);
+                            // TODO: queue a task to restart animation, if restart-animation is set
+                            return
                         }
+                        self.init_pending_request(&url, &src);
+                        self.image_request.set(ImageRequestPhase::Pending);
+                        self.fetch_image(&url);
+                    },
+                    (_, State::Broken) | (_, State::Unavailable) => {
+                        // Step 12.5
+                        current_request.parsed_url = Some(url.clone());
+                        current_request.source_url = Some(src.clone());
+                        current_request.image = None;
+                        current_request.metadata = None;
+                        let document = document_from_node(self);
+                        current_request.blocker = Some(LoadBlocker::new(&*document, LoadType::Image(url.clone())));
+                        self.fetch_image(&url);
+                    },
+                    (_, _) => {
+                        // step 12.6
+                        self.init_pending_request(&url, &src);
+                        self.image_request.set(ImageRequestPhase::Pending);
+                        self.fetch_image(&url);
+                    },
+                }
+            }
+        }
+    }
+
+    /// Step 8-12 of html.spec.whatwg.org/multipage/#update-the-image-data
+    fn update_the_image_data_sync_steps(&self) {
+        let document = document_from_node(self);
+        // Step 8
+        // TODO: take pixel density into account
+        match self.select_image_source() {
+            Some(src) => {
+                // Step 10
+                self.dispatch_progress_event(atom!("loadstart"), 0, None);
```

Comment:
`dispatch_progress_event` dispatches a progress event via the scheduling of a task, is that what you mean? Should I change the method name to better reflect this?